src/backend/binaryen/utils.ts

```diff
@@ -1587,123 +1673,104 @@ export namespace FunctionalFuncs {
-        if (charCode <= 0x7f) {
-            utf8String += String.fromCharCode(charCode);
-        } else if (charCode <= 0x7ff) {
-            utf8String += String.fromCharCode(0xc0 | ((charCode >> 6) & 0x1f));
-            utf8String += String.fromCharCode(0x80 | (charCode & 0x3f));
-        } else {
-            utf8String += String.fromCharCode(0xe0 | ((charCode >> 12) & 0x0f));
-            utf8String += String.fromCharCode(0x80 | ((charCode >> 6) & 0x3f));
-            utf8String += String.fromCharCode(0x80 | (charCode & 0x3f));
-        }
+    export function getWasmStructFieldByIndex(
+        module: binaryen.Module,
+        ref: binaryen.ExpressionRef,
+        typeRef: binaryen.Type,
+        idx: number,
+    ) {
+        return binaryenCAPI._BinaryenStructGet(
+            module.ptr,
+            idx,
+            ref,
+            typeRef,
+            false,
+        );
     }
 
-    return utf8String;
-}
-
-export function clearWasmStringMap() {
-    wasmStringMap.clear();
-}
-
-export const enum StructFieldIndex {
-    VTABLE_INDEX = 0,
-}
-
-export const enum VtableFieldIndex {
-    META_INDEX = 0,
-}
-
-export const SIZE_OF_META_FIELD = 12;
-
-export const enum MetaDataOffset {
-    TYPE_ID_OFFSET = 0,
-    IMPL_ID_OFFSET = 4,
-    COUNT_OFFSET = 8,
-    FIELDS_PTR_OFFSET = 12,
-}
+    export function getWASMObjectVtable(
+        module: binaryen.Module,
+        ref: binaryen.ExpressionRef,
+    ) {
+        return getWasmStructFieldByIndex(
+            module,
+            ref,
+            baseVtableType.typeRef,
+            StructFieldIndex.VTABLE_INDEX,
+        );
+    }
 
-export const enum MetaFieldOffset {
-    NAME_OFFSET = 0,
-    FLAG_AND_INDEX_OFFSET = 4,
-    TYPE_OFFSET = 8,
-}
+    export function getWASMObjectMeta(
+        module: binaryen.Module,
+        ref: binaryen.ExpressionRef,
+    ) {
+        const vtable = getWASMObjectVtable(module, ref);
+        return getWasmStructFieldByIndex(
+            module,
+            vtable,
+            binaryen.i32,
+            VtableFieldIndex.META_INDEX,
+        );
+    }
 
-export function getFieldFromMetaByOffset(
-    module: binaryen.Module,
-    meta: binaryen.ExpressionRef,
-    offset: number,
-) {
-    return module.i32.load(offset, memoryAlignment, meta);
-}
+    export function isUndefinedIndex(
+        module: binaryen.Module,
+        indexRef: binaryen.ExpressionRef,
+    ) {
+        return module.i32.eq(indexRef, module.i32.const(268435455));
+    }
 
-export function getWasmStructFieldByIndex(
-    module: binaryen.Module,
-    ref: binaryen.ExpressionRef,
-    typeRef: binaryen.Type,
-    idx: number,
-) {
-    return binaryenCAPI._BinaryenStructGet(
-        module.ptr,
-        idx,
-        ref,
-        typeRef,
-        false,
-    );
-}
+    export function isUndefinedFlag(
+        module: binaryen.Module,
+        flagRef: binaryen.ExpressionRef,
+    ) {
+        return module.i32.eq(flagRef, module.i32.const(15));
```

Comment:
What does 15 mean?

Reply:
In function `find_property_flag_and_index`, we return `-1` to represent that we don't found the property name in meta, but the `-1` represent that both flag and index, we calculate `-1 & META_FLAG_MASK`, then we will get `15`, so here I compare flagRef and 15 to determine if we found the property name.

Reply:
Can we use an expression to calculate the value? Or at least leave a comment here to indicate what does 15 mean